src/Client/V2/Service/MessageHandler.cs

```diff
@@ -0,0 +1,313 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Newtonsoft.Json.Linq;
+
+namespace EmsApi.Client.V2
+{
+    /// <summary>
+    /// Handles authentication and setting other EMS API specific headers. This class will handle
+    /// requesting authentication tokens when necessary.
+    /// </summary>
+    /// <remarks>
+    /// Because authentication is not attempted until the first time the service is accessed,
+    /// we provide a callback for authentication errors instead of throwing an exception, since
+    /// it can come at an unexpected time for the client.
+    /// </remarks>
+    internal class MessageHandler : DelegatingHandler
+    {
+        private EmsApiServiceConfiguration m_serviceConfig;
+
+        // We store copies of some service config variables to determine if a change to the service
+        // configuration affects us.
+        private string m_endpoint, m_userName, m_pass, m_clientId, m_clientSecret, m_trustedAuthName;
+
+        /// <summary>
+        /// The cache of tokens we have retrieved.
+        /// The keys to this dictionary are defined by <seealso cref="AuthConfig"/>.
+        /// Note: Tokens in the cache may have expired.
+        /// </summary>
+        private readonly ConcurrentDictionary<string, AuthToken> m_tokens = new ConcurrentDictionary<string, AuthToken>();
+
+        /// <summary>
+        /// Sets the current service configuration, causing the authentication
+        /// to become invalid if the endpoint, username, or password changed.
+        /// </summary>
+        public EmsApiServiceConfiguration ServiceConfig
+        {
+            set
+            {
+                bool deAuth = false;
+                if( value.Endpoint != m_endpoint )
+                {
+                    m_endpoint = value.Endpoint;
+                    deAuth = true;
+                }
+                if( value.UserName != m_userName )
+                {
+                    m_userName = value.UserName;
+                    deAuth = true;
+                }
+                if( value.Password != m_pass )
+                {
+                    m_pass = value.Password;
+                    deAuth = true;
+                }
+                if( value.ApiClientId != m_clientId )
+                {
+                    m_clientId = value.ApiClientId;
+                    deAuth = true;
+                }
+                if( value.ApiClientSecret != m_clientSecret )
+                {
+                    m_clientSecret = value.ApiClientSecret;
+                    deAuth = true;
+                }
+                if( value.TrustedAuthName != m_trustedAuthName )
+                {
+                    m_trustedAuthName = value.TrustedAuthName;
+                    deAuth = true;
+                }
+
+                if( deAuth )
+                    InvalidateAuthentication();
```

Comment:
I think you can replace this call with `ClearAuthenticationCache()` instead and delete `InvalidateAuthentication`.

Reply:
Good catch!